# Segmenting and Clustering Neighborhoods in the city of Toronto, Canada

## Part 3 week 3 IBM Data Science - Applied Data Science Capstone

In Part 1 a dataframe with the Postal Code of each canadian borough and neighborhood was developed. Part 2 consisted in adding the latitude and the longitude coordinates of all neighborhoods. Part 3 consists in exploring and clustering the neighborhoods in Toronto, Canada. It will include maps to visualize neighborhoods in Toronto and how they cluster together.

### **PART 1**

**1. Installing and importing libraries**

In [1]:
# This code installs the required libraries
!pip install beautifulsoup4
!conda install -c conda-forge geopy --yes
!conda install -c conda-forge folium=0.5.0 --yes

print('Libraries installed successfully!')

Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

# All requested packages already installed.

Libraries installed successfully!


In [2]:
# This code imports the required libraries
import requests
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup

print('Libraries imported successfully!')

Libraries imported successfully!


**2. Scraping the list of postal codes in Canada dataset from Wikipedia**

In [3]:
# This code reads the Wikipedia website content of the server's response
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
results = requests.get(url).text

# This code parses the html document from the Wikipedia website using BeautifulSoup library
Canada_data = BeautifulSoup(results, 'html.parser')
wikipedia_table = Canada_data.find('table')

# This code converts the Wikipedia html table into a DataFrame using Pandas library
column_names = ['Postal Code', 'Borough', 'Neighborhood']
df = pd.DataFrame(columns=column_names)
df.head()

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

In [4]:
# This code searches all the postcodes, boroughs, neighborhoods available
for tr_cell in wikipedia_table.find_all('tr'):
    row_data = []
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data) == 3:
        df.loc[len(df)] = row_data
    
# This code displays the 12 first results in the df DataFrame
df.head(12)

Postal Code           Borough                                 Neighborhood
0          M1A      Not assigned                                 Not assigned
1          M2A      Not assigned                                 Not assigned
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7          M8A      Not assigned                                 Not assigned
8          M9A         Etobicoke      Islington Avenue, Humber Valley Village
9          M1B       Scarborough                               Malvern, Rouge
10         M2B      Not assigned                                 Not assigned
11         M3B        North York                                    Don Mills

**3. Data cleaning**

In [5]:
# This code removes boroughs with "not assigned" values
df = df[df['Borough'].str.contains("Not assigned") == False].reset_index()

# This code displays the 12 first results in the df DataFrame
df.head(12)

index Postal Code           Borough  \
0       2         M3A        North York   
1       3         M4A        North York   
2       4         M5A  Downtown Toronto   
3       5         M6A        North York   
4       6         M7A  Downtown Toronto   
5       8         M9A         Etobicoke   
6       9         M1B       Scarborough   
7      11         M3B        North York   
8      12         M4B         East York   
9      13         M5B  Downtown Toronto   
10     14         M6B        North York   
11     17         M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [6]:
# This code removes the first column
df.drop(['index'], axis = 1, inplace = True)

# This code displays the 12 first results in the cleaned df DataFrame
df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [7]:
# This code prints the dimensions of the df DataFrame
print('The shape for df DataFrame is:', df.shape)

The shape for df DataFrame is: (103, 3)


The dataframe contains 103 Postal Codes (rows) and 3 columns: Postal Code, Borough, Neighborhood

### **PART 2**

**4. Adding the latitude and longitude coordinates to the DataFrame**

In [8]:
# This code defines a function to get the coordinates (latitude, longitude) for all neighborhoods
def get_geocode(postal_code):
    
    # initialize the variable to None
    lat_lng_coords = None
    
    # loop until the coordinates are obtained
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
        
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    return latitude, longitude

In [9]:
# This code reads the geographical coordinates of each postal code using pandas library
locgeo_df = pd.read_csv('https://cocl.us/Geospatial_data')

# This code displays the 12 first results of the locgeo_df DataFrame
locgeo_df.head(12)

Postal Code   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849

The shape of the 'locgeo_df' DataFrame should be checked before joining 'df' and 'locgeo_df' DataFrames to ensure both have the same shape.

In [10]:
# This code prints the dimensions of the locgeo_df DataFrame
print('The shape for locgeo_df DataFrame is:', locgeo_df.shape)

The shape for locgeo_df DataFrame is: (103, 3)


The shape of 'df' and 'locgeo_df' DataFrames is the same so Latitude and Longitude columns from the 'locgeo_df' DataFrame can be added to the existing 'df' DataFrame 

In [11]:
# This code joins df and locgeo_df DataFrames on Postal Code column
df = df.join(locgeo_df.set_index('Postal Code'), on = 'Postal Code')

# This code displays the 12 first results of the new df DataFrame
df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

### **PART 3**

**5. Using Foursquare API to explore the neighborhoods and segment them**

In [12]:
# This code installs geocoder library
!conda install -c conda-forge geocoder --yes

# This code imports the required libraries
import geocoder
from geopy.geocoders import Nominatim # Convert an address into latitude and longitude values

import matplotlib.cm as cm # Matplotlib and associated plotting modules
import matplotlib.colors as colors # Matplotlib and associated plotting modules

from sklearn.cluster import KMeans # import k-means from clustering stage

import folium # map rendering library

print('Libraries installed and imported successfully!')

Solving environment: ...working... done

# All requested packages already installed.

Libraries installed and imported successfully!


Let's get the geographical coordinates of Toronto, Canada

In [13]:
# This code gets the geographical coordinates of Toronto
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent = "toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geographical coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the neighborhoods in Toronto

In [14]:
# Create a map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start = 11)

# Add markers to Toronto map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat, lng],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    ).add_to(map_toronto)
    
map_toronto

Defining Foursquare Credentials and Version

In [15]:
# This code defines Foursquare credentials and version
CLIENT_ID = 'SF3QURMDBEA11X4EDFJPY30R5AYACPJBBHL0I1EAVPYLGJC0' # Foursquare ID
CLIENT_SECRET = 'O5PEG4DFUJKZW3RNE4E2WSSJHUSVCIWPF3BQUNQJHIHZLUNJ' # Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET' + CLIENT_SECRET)

Your credentials:
CLIENT_ID: SF3QURMDBEA11X4EDFJPY30R5AYACPJBBHL0I1EAVPYLGJC0
CLIENT_SECRETO5PEG4DFUJKZW3RNE4E2WSSJHUSVCIWPF3BQUNQJHIHZLUNJ


**6. Exploring Neighborhoods in Toronto, Canada**

Let's create a function to repeat the same process to all neighborhoods in Toronto, Canada

In [16]:
# This code explores the neighborhoods in Toronto, Canada
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID,
            CLIENT_SECRET,
            VERSION,
            lat,
            lng,
            radius,
            LIMIT)
        
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name,
            lat,
            lng,
            v['venue']['name'],
            v['venue']['location']['lat'],
            v['venue']['location']['lng'],
            v['venue']['categories'][0]['name']) for v in results])
        
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood',
                            'Neighborhood Latitude',
                            'Neighborhood Longitude',
                            'Venue',
                            'Venue Latitude',
                            'Venue Longitude',
                            'Venue Category']
    
    return(nearby_venues)

In [17]:
# This code returns a list of venues near the current location
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                latitudes=df['Latitude'],
                                longitudes=df['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Let's check the size of the resulting DataFrame

In [18]:
# This code prints the first 12 nearby venues in Toronto neighborhoods
toronto_venues.head(12)

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                   Parkwoods              43.753259              -79.329656   
1                   Parkwoods              43.753259              -79.329656   
2                   Parkwoods              43.753259              -79.329656   
3                   Parkwoods              43.753259              -79.329656   
4            Victoria Village              43.725882              -79.315572   
5            Victoria Village              43.725882              -79.315572   
6            Victoria Village              43.725882              -79.315572   
7            Victoria Village              43.725882              -79.315572   
8   Regent Park, Harbourfront              43.654260              -79.360636   
9   Regent Park, Harbourfront              43.654260              -79.360636   
10  Regent Park, Harbourfront              43.654260              -79.360636   
11  Regent Park, Harbourfront              43.654260              -79.360636   

                                Venue  Venue Latitude  Venue Longitude  \
0                     Brookbanks Park       43.751976       -79.332140   
1                     Brookbanks Pool       43.751389       -79.332184   
2                       Variety Store       43.751974       -79.333114   
3   Corrosion Service Company Limited       43.752432       -79.334661   
4              Victoria Village Arena       43.723481       -79.315635   
5                           Portugril       43.725819       -79.312785   
6                         Tim Hortons       43.725517       -79.313103   
7                          Pizza Nova       43.725824       -79.312860   
8                    Roselle Desserts       43.653447       -79.362017   
9                       Tandem Coffee       43.653559       -79.361809   
10             Cooper Koo Family YMCA       43.653249       -79.358008   
11                Body Blitz Spa East       43.654735       -79.359874   

                Venue Category  
0                         Park  
1                         Pool  
2            Food & Drink Shop  
3   Construction & Landscaping  
4                 Hockey Arena  
5        Portuguese Restaurant  
6                  Coffee Shop  
7                  Pizza Place  
8                       Bakery  
9                  Coffee Shop  
10         Distribution Center  
11                         Spa

In [19]:
# This code prints the dimensions of the toronto_venues DataFrame
toronto_venues.shape

(2124, 7)

Let's check how many venues were returned for each neighborhood

In [20]:
# This code counts the number of venues in each neighborhood
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            5   
Alderwood, Long Branch                                              10   
Bathurst Manor, Wilson Heights, Downsview North                     22   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   23   
...                                                                ...   
Willowdale, Willowdale East                                         34   
Willowdale, Willowdale West                                          5   
Woburn                                                               5   
Woodbine Heights                                                     5   
York Mills West                                                      3   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             5   
Alderwood, Long Branch                                               10   
Bathurst Manor, Wilson Heights, Downsview North                      22   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    23   
...                                                                 ...   
Willowdale, Willowdale East                                          34   
Willowdale, Willowdale West                                           5   
Woburn                                                                5   
Woodbine Heights                                                      5   
York Mills West                                                       3   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            5               5   
Alderwood, Long Branch                              10              10   
Bathurst Manor, Wilson Heights, Downsview North     22              22   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   23              23   
...                                                ...             ...   
Willowdale, Willowdale East                         34              34   
Willowdale, Willowdale West                          5               5   
Woburn                                               5               5   
Woodbine Heights                                     5               5   
York Mills West                                      3               3   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      5   
Alderwood, Long Branch                                        10   
Bathurst Manor, Wilson Heights, Downsview North               22   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             23   
...                                                          ...   
Willowdale, Willowdale East                                   34   
Willowdale, Willowdale West                                    5   
Woburn                                                         5   
Woodbine Heights                                               5   
York Mills West                                                3   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     5  
Alderwood, Long Branch                                       10 

**7. Analyze Each Neighborhood**

In [21]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to DataFrame
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  ...  Trail  Train Station  Truck Stop  \
0                    0             0  ...      0              0           0   
1                    0             0  ...      0              0           0   
2                    0             0  ...      0              0           0   
3                    0             0  ...      0              0           0   
4                    0             0  ...      0              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  
0                0         0            0              0  
1                0         0            0              0  
2                0         0            0              0  
3                0         0            0              0  
4                0         0            0              0  

[5 rows x 274 columns]

And let's examine the new DataFrame size

In [22]:
# This code gets the dimensions of the toronto_onehot DataFrame
toronto_onehot.shape

(2124, 274)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [23]:
# This code groups rows by neighborhood
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                0.0             0.0      0.0                 0.0   
1                0.0             0.0      0.0                 0.0   
2                0.0             0.0      0.0                 0.0   
3                0.0             0.0      0.0                 0.0   
4                0.0             0.0      0.0                 0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.043478   

   ...  Trail  Train Station  Truck Stop  Vegetarian / Vegan Restaurant  \
0  ...    0.0            0.0         0.0                            0.0   
1  ...    0.0            0.0         0.0                            0.0   
2  ...    0.0            0.0         0.0                            0.0   
3  ...    0.0            0.0         0.0                            0.0   
4  ...    0.0            0.0         0.0                            0.0   

   Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0               0.0                    0.0              0.0       0.0   
1               0.0                    0.0              0.0       0.0   
2               0.0                    0.0              0.0       0.0   
3               0.0                    0.0              0.0       0.0   
4               0.0                    0.0              0.0       0.0   

   Wings Joint  Women's Store  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0  

[5 rows x 274 columns]

Let's confirm the new size

In [24]:
# This code gets the dimensions of the toronto_grouped DataFrame
toronto_grouped.shape

(96, 274)

Let's print each neighborhood along with the top 10 most common venues and write a function to sort the venues in descending order

In [25]:
# This function sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new DataFrame and display the top 10 venues for each neighborhood

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new DataFrame
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)
    
neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge          Skating Rink        Breakfast Spot   
1        Sandwich Place          Skating Rink                   Gym   
2                  Bank    Frozen Yogurt Shop           Bridal Shop   
3                  Bank    Chinese Restaurant   Japanese Restaurant   
4    Italian Restaurant           Coffee Shop              Pharmacy   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store         Women's Store             Drugstore   
1           Coffee Shop              Pharmacy          Dance Studio   
2        Sandwich Place                 Diner            Restaurant   
3         Women's Store   Distribution Center               Dog Run   
4     Indian Restaurant                   Pub      Sushi Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Distribution Center               Dog Run       Doner Restaurant  
1    Athletics & Sports                   Pub          Women's Store  
2         Deli / Bodega           Supermarket       Sushi Restaurant  
3      Doner Restaurant            Donut Shop              Drugstore  
4               Butcher            Restaurant   Fast Food Restaurant

**8. Cluster Neighborhoods**

Running k-means to cluster the neighborhood into 5 clusters

In [27]:
# Setting the number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# Running k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# Checking cluster labels generated for each row in the DataFrame
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Let's create a new DataFrame that includes the cluster as well as the top 10 venues for each neighborhood

In [28]:
# Adding clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df

# Merging toronto_grouped with df to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

# Checking the last columns
toronto_merged.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             0.0          Hockey Arena   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

    2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0       Food & Drink Shop  Construction & Landscaping                  Pool   
1   Portuguese Restaurant                 Pizza Place           Coffee Shop   
2                    Park                      Bakery        Breakfast Spot   
3  Furniture / Home Store                 Coffee Shop    Athletics & Sports   
4                   Diner            Sushi Restaurant           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0            Donut Shop                 Diner         Discount Store   
1                 Diner        Discount Store    Distribution Center   
2                  Café                   Pub                Theater   
3             Gift Shop              Boutique  Vietnamese Restaurant   
4                   Bar              Beer Bar          Smoothie Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Distribution Center               Dog Run       Doner Restaurant  
1                Dog Run      Doner Restaurant             Donut Shop  
2  Performing Arts Venue            Shoe Store                Brewery  
3      Accessories Store  Ethiopian Restaurant            Escape Room  
4         Sandwich Place         Burrito Place             Restaurant

Finally, let's visualize the resulting clusters

In [29]:
# Creating map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# Setting color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

# Adding markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
    
map_clusters

**9. Examining Clusters**

Each cluster will be examined and determined the discriminating venue categories that distinguish each cluster.

Cluster 1

In [30]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0          North York             0.0                  Park   
1          North York             0.0          Hockey Arena   
2    Downtown Toronto             0.0           Coffee Shop   
3          North York             0.0        Clothing Store   
4    Downtown Toronto             0.0           Coffee Shop   
..                ...             ...                   ...   
97   Downtown Toronto             0.0           Coffee Shop   
98          Etobicoke             0.0                 River   
99   Downtown Toronto             0.0           Coffee Shop   
100      East Toronto             0.0           Yoga Studio   
102         Etobicoke             0.0                   Gym   

      2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0         Food & Drink Shop  Construction & Landscaping                  Pool   
1     Portuguese Restaurant                 Pizza Place           Coffee Shop   
2                      Park                      Bakery        Breakfast Spot   
3    Furniture / Home Store                 Coffee Shop    Athletics & Sports   
4                     Diner            Sushi Restaurant           Yoga Studio   
..                      ...                         ...                   ...   
97                     Café                       Hotel            Restaurant   
98                     Pool        Ethiopian Restaurant           Escape Room   
99      Japanese Restaurant            Sushi Restaurant               Gay Bar   
100           Auto Workshop        Gym / Fitness Center         Garden Center   
102              Kids Store              Discount Store         Tanning Salon   

    5th Most Common Venue        6th Most Common Venue  7th Most Common Venue  \
0              Donut Shop                        Diner         Discount Store   
1                   Diner               Discount Store    Distribution Center   
2                    Café                          Pub                Theater   
3               Gift Shop                     Boutique  Vietnamese Restaurant   
4                     Bar                     Beer Bar          Smoothie Shop   
..                    ...                          ...                    ...   
97    Japanese Restaurant                          Gym             Steakhouse   
98      Electronics Store  Eastern European Restaurant    Dumpling Restaurant   
99             Restaurant                  Men's Store                   Café   
100                Garden         Fast Food Restaurant         Farmers Market   
102         Burrito Place               Sandwich Place           Burger Joint   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Distribution Center               Dog Run       Doner Restaurant  
1                  Dog Run      Doner Restaurant             Donut Shop  
2    Performing Arts Venue            Shoe Store                Brewery  
3        Accessories Store  Ethiopian Restaurant            Escape Room  
4           Sandwich Place         Burrito Place             Restaurant  
..                     ...                   ...                    ...  
97             Salad Place   American Restaurant       Asian Restaurant  
98               Drugstore      Department Store             Donut Shop  
99                     Pub           Yoga Studio                  Hotel  
100     Light Rail Station            Comic Shop            Pizza Place  
102                 Bakery       Supplement Shop      Convenience Store  

[90 rows x 12 columns]

Cluster 2

In [31]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
21        York             1.0                  Park                  Pool   
35   East York             1.0                  Park     Convenience Store   
52  North York             1.0                  Park             Drugstore   
64        York             1.0                  Park             Drugstore   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
21         Women's Store      Greek Restaurant         Deli / Bodega   
35   Dumpling Restaurant        Discount Store   Distribution Center   
52                 Diner        Discount Store   Distribution Center   
64                 Diner        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
21           Escape Room     Electronics Store  Eastern European Restaurant   
35               Dog Run      Doner Restaurant                   Donut Shop   
52               Dog Run      Doner Restaurant                   Donut Shop   
64               Dog Run      Doner Restaurant                   Donut Shop   

   9th Most Common Venue       10th Most Common Venue  
21   Dumpling Restaurant                    Drugstore  
35             Drugstore  Eastern European Restaurant  
52   Dumpling Restaurant              College Stadium  
64   Dumpling Restaurant              College Stadium

Cluster 3

In [32]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
32       Scarborough             2.0            Playground   
85       Scarborough             2.0                  Park   
91  Downtown Toronto             2.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32         Women's Store             Drugstore                 Diner   
85          Intersection            Playground            Donut Shop   
91            Playground                 Trail  Ethiopian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32        Discount Store   Distribution Center               Dog Run   
85                 Diner        Discount Store   Distribution Center   
91           Escape Room     Electronics Store           Event Space   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32             Doner Restaurant            Donut Shop    Dumpling Restaurant  
85                      Dog Run      Doner Restaurant              Drugstore  
91  Eastern European Restaurant   Dumpling Restaurant       Department Store

Cluster 4

In [33]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
57   North York             3.0  Paper / Office Supplies Store   
101   Etobicoke             3.0                 Baseball Field   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57         Baseball Field         Women's Store             Drugstore   
101         Women's Store        Discount Store   Distribution Center   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
57         Discount Store   Distribution Center               Dog Run   
101               Dog Run      Doner Restaurant            Donut Shop   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
57       Doner Restaurant            Donut Shop    Dumpling Restaurant  
101             Drugstore   Dumpling Restaurant   Fast Food Restaurant

Cluster 5

In [34]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12  Scarborough             4.0                   Bar         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12             Drugstore        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
12               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
12   Dumpling Restaurant   Fast Food Restaurant